### PR TITLE
Added support for OBJECT IDENTIFIER types

### PR DESCRIPTION
**Pull request by ecarson**
- Added support for Object Identifier types specificed by either dot
   notation or MIB string.
 - These are currently non updateable

### CODE REVIEW

File: netsnmpagent.py [hunk 1]
```diff
@@ -679,6 +692,83 @@ def DisplayString(self, initval = None, oidstr = None, writable = True, context
...
+					agent._objs[context][oidstr] = self
+
+			def _set_oid_value(self, oid_value):
+				if oid_value is not None:
+					oid, oid_len = agent.determine_oid_and_length(oid_value)
+					self._object_id = oid
+					self._cvar      = self._object_id
+					self._data_size = oid_len.value * ctypes.sizeof(c_oid)
+				else:
+					self._object_id = (c_oid * 0)()
```
Comment (ian-h-chamberlain) [inline netsnmpagent.py:755]: What does `* 0` mean here? Is this something similar to `[0]` to dereference a C pointer?

File: netsnmpagent.py [hunk 2]
```diff
@@ -679,6 +692,83 @@ def DisplayString(self, initval = None, oidstr = None, writable = True, context
...
+					self._cvar      = self._object_id
+					self._data_size = 0
+
+			def value(self):
+				return "." + ".".join([str(oid) for oid in self._object_id])
+
+			def cref(self, **kwargs):
```
Comment (ian-h-chamberlain) [inline netsnmpagent.py]: `**kwargs` unused here could be removed?

File: netsnmpagent.py [hunk 3]
```diff
@@ -679,6 +692,83 @@ def DisplayString(self, initval = None, oidstr = None, writable = True, context
...
+					if self._callback_handler is not None:
+						_inject_custom_handler(self._callback_handler, handler_reginfo)
+
+					# Finally, we keep track of all registered SNMP objects for the
+					# getRegistered() method.
+					agent._objs[context][oidstr] = self
+
+			def _set_oid_value(self, oid_value):
+				if oid_value is not None:
+					oid, oid_len = agent.determine_oid_and_length(oid_value)
```
Comment (ian-h-chamberlain) [inline netsnmpagent.py:750]: Does `agent.determine_oid_and_length(oid_value)` resolve and return the `.`-separated OID for a string-formatted human readable OID? Will we raise errors there if no such OID exists?

File: netsnmpagent.py [hunk 4]
```diff
@@ -400,12 +399,26 @@ def _prepareRegistration(self, oidstr, writable = True):
...
+
+	def _prepareRegistration(self, oidstr, writable = True):
+		""" Prepares the registration of an SNMP object.
+
+		    "oidstr" is the OID to register the object at.
+		    "writable" indicates whether "snmpset" is allowed. """
+
+		# Make sure the agent has not been start()ed yet
+		if self._status != netsnmpAgentStatus.REGISTRATION:
+			raise netsnmpAgentException("Attempt to register SNMP object " \
```
Comment (ian-h-chamberlain) [inline netsnmpagent.py]: I think the trailing `\` here is unnecessary since it's within parentheses

File: netsnmpagent.py (hunk 3 above)
Comment (ian-h-chamberlain) [reply netsnmpagent.py:750]: Ah, nvm, I just read this PR in the wrong order. I see what's going on now

**Review by ian-h-chamberlain — APPROVED** [review]
Have you tested this against python2? I guess we're basically maintaining a full fork at this point but it does need to be Py2 compatible if we want to merge this upstream.

File: netsnmpagent.py (hunk 1 above)
Comment (ecarson) [reply netsnmpagent.py:755]: It's the weird array syntax for c_types:
https://docs.python.org/3/library/ctypes.html#arrays

This specifies a 0 length array (ie: nothing).

File: netsnmpagent.py (hunk 2 above)
Comment (ecarson) [reply netsnmpagent.py]: Will do

File: netsnmpagent.py (hunk 4 above)
Comment (ecarson) [reply netsnmpagent.py]: Will remove

**Pull request merged**